src/ActiveLogin.Authentication.BankId.AspNetCore.UAParser/UAParserDeviceDetector.cs

```diff
@@ -0,0 +1,156 @@
+        {
+            var hasMajor = int.TryParse(clientInfo.OS.Major, out var major);
+            var hasMinor = int.TryParse(clientInfo.OS.Minor, out var minor);
+            var hasPatch = int.TryParse(clientInfo.OS.Patch, out var patch);
+
+            return (hasMajor, hasMinor, hasPatch) switch
+            {
+                (true, true, true) => new BankIdSupportedDeviceOsVersion(major, minor, patch),
+                (true, true, false) => new BankIdSupportedDeviceOsVersion(major, minor),
+                (true, false, false) => new BankIdSupportedDeviceOsVersion(major),
+                _ => BankIdSupportedDeviceOsVersion.Empty
+            };
+        }
+
+        private BankIdSupportedDeviceType GetDeviceType(ClientInfo clientInfo)
+        {
+            var deviceOs = GetDeviceOs(clientInfo);
+            var deviceType = GetDeviceType(deviceOs);
+            var isMobileBrowser = IsMobileBrowser(clientInfo.UA);
+
+            return (deviceType, isMobileBrowser) switch
+            {
+                (BankIdSupportedDeviceType.Unknown, true) => BankIdSupportedDeviceType.Mobile,
+                _ => deviceType
+            };
+        }
+
+        private BankIdSupportedDeviceType GetDeviceType(BankIdSupportedDeviceOs deviceOs)
+        {
+            return deviceOs switch
+            {
+                BankIdSupportedDeviceOs.Windows => BankIdSupportedDeviceType.Desktop,
+                BankIdSupportedDeviceOs.MacOs => BankIdSupportedDeviceType.Desktop,
+
+                BankIdSupportedDeviceOs.Ios => BankIdSupportedDeviceType.Mobile,
+                BankIdSupportedDeviceOs.Android => BankIdSupportedDeviceType.Mobile,
+                BankIdSupportedDeviceOs.WindowsPhone => BankIdSupportedDeviceType.Mobile,
+
+                _ => BankIdSupportedDeviceType.Unknown
+            };
+        }
+
+        private bool IsChrome(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "chrome");
+        }
+
+        private bool IsSafari(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "safari");
+        }
+
+        private bool IsEdge(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "edge");
+        }
+
+        private bool IsFirefox(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "firefox");
+        }
+
+        private bool IsSamsungBrowser(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "samsung");
+        }
+
+        private bool IsOpera(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "opera");
+        }
+
+        private bool IsMobileBrowser(UserAgent userAgent)
+        {
+            return IsBrowser(userAgent, "mobile");
+        }
+
+        private bool IsBrowser(UserAgent userAgent, string browser)
+        {
+            return userAgent.Family.ToLower().Contains(browser.ToLower(), StringComparison.InvariantCulture);
```

Comment:
We've done containts before, but maybe we can be more exact in our match now? Not sure in what format UA parser returns it. Could be do (for example) `StartsWith`?

Also, I think we could do `StringComparer.InvariantCultureIgnoreCase` and skip `ToLower()`.

The same applies to `IsOs()` below.

Reply:
Replaced ToLower() with StringComparer.InvariantCultureIgnoreCase.

Reply:
Nice! Did you see if there was any way to narrow it down to (something like StartsWith) instead of Contains?

Reply:
I have noticed that the string in ClientInfo.UA.Family could be e.g. "Mobile Safari" or "Chrome Mobile" i.e. it does not always start with the name of the browser.